### PR TITLE
[CoSim] EMPIRE replacement

### PR DESCRIPTION
**Pull request by philbucher**
This PR aims at replacing EMPIRE-usages with CoSimulation
For this the functions defined in the `EMPIRE_API.h`were reimplemented inside `co_sim_EMPIRE_API.h`, now being based on **FileIO** (instead of MPI-Ports, which is very problematic especially on newer cluster systems).
Features:
- `EMPIRE_API.h` can be replaced **plug&play** (I tried this already) with `co_sim_EMPIRE_API.h` => codes that have an interface to EMPIRE will be able to use this to couple to CoSimulation instead without changing anything!
- `co_sim_EMPIRE_API.h` is header-only (one file) => external codes don't have to link against anything any more
- The mesh-exchange is based on vtk (so far only ASCII), hence the mesh can directly be visualized in Paraview
- `co_sim_EMPIRE_API.h` is exposed to python as a submodule of the CoSimApp and has a very convenient interface to Kratos/ModelPart
- the integration/usage in the CoSimApp-workflow will be the next step after this is merged
- Tests for all functionalities were added ofc

Note that this is only required for backwards compatibility with codes that we already work with, the proper CoSim-FileIO will come in the future

@joris13 @navaneethkn you can use this directly to couple with OpenFOAM, since it has an EMPIRE-interface (developed here at the chair). Let us know if you are interested

part of #5140

### CODE REVIEW

File: applications/CoSimulationApplication/custom_io/co_sim_EMPIRE_API.h [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+{
+    std::string current_line_substr = rCurrentLine.substr(rCurrentLine.find(rKeyWord) + rKeyWord.length());
+    std::istringstream line_stream(current_line_substr);
+    line_stream >> rNumber;
+}
+
+} // namespace EMPIRE_API_helpers
+
+/***********************************************************************************************
+ * \brief Establishes the necessary connection with the Emperor
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/custom_io/co_sim_EMPIRE_API.h:177]:
```suggestion
 * \brief Does necessary initial tasks of creating and organizing the i/o files.
```

**Comment by philbucher** [comment]
I added a "test" for having the api-header included in multiple files. This fails with the same error as in FLOWer if sth is not done right.

File: applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        }
+        if (UseConditions) {
+            rModelPart.CreateNewCondition(condition_name_map.at(num_nodes_elem), i+1, elem_node_ids, p_props);
+        } else {
+            rModelPart.CreateNewElement(element_name_map.at(num_nodes_elem), i+1, elem_node_ids, p_props);
+        }
+    }
+
+    // deallocating memory
+    // TODO what abt the memory that was allocated inside "EMPIRE_API_recvMesh"?
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp]: Below you are deleting the memory which is allocated inside the `EMPIRE_API_recvMesh`.  You are not allocating any memory here !

File: applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    delete [] nodes;
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp:314]: these are double pointers ! 
for a 2 nodes you should do something like 
```cpp
for i : {1,2}
  delete [] nodes[i]
```

if you just say delete [] nodes you are just deleting the memory for the vector of pointers which are pointing to the nodal coordinates but not the memory for the coordinates themselves.

File: applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+conv_signal_file_name = "EMPIRE_convergence_signal.dat" # this is hardcoded in C++
+
+def GetFilePath(fileName):
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), fileName)
+
+class TestCoSim_EMPIRE_API(KratosUnittest.TestCase):
+
+    def test_unused_fcts(self):
+        KratosCoSim.EMPIRE_API.EMPIRE_API_Connect("dummy.xml")
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py:19]: This is not an actual xml anymore right ?

File: applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+        array = [1.0, 2.5, 3.5, -99.11, -0.02, 555.5, -99.114, -99.11, -0.02, 555.5, 0.02, 565.5, 10.0, 78.44]
+
+        self.__TestArrayReceive(fct_to_test, file_name_fct_ptr, array)
+
+    def test_EMPIRE_API_sendSignal_double(self):
+        fct_to_test = KratosCoSim.EMPIRE_API.EMPIRE_API_sendSignal_double
+        file_name_fct_ptr = GetSignalFileName
+
+        array = [1.0, 2.5, 3.5, -99.11, -0.02, 555.5, -99.114, 0.02, 565.5, 10.0, 78.44]
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py:131]: Is this required ? The above test also tests the same !

Comment (philbucher) [reply applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py:131]: the test above tests `EMPIRE_API_sendDataField`, this test is for `EMPIRE_API_sendSignal_double`
I just set it up to use the same testing framework

Comment (adityaghantasala) [reply applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py:131]: ah ok .. i missed line 128 !

File: applications/CoSimulationApplication/tests/cpp_tests/test_multiple_includes_api_headers.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+This file includes the headers that are also included in other codes.
+The reason is to test if the corresponding header can be included in multiple files without causing problems
+(e.g. causing linking-erros if free functions are not static)
+The "other" includ is supposed to be in the "add_xxx_to_python"
+
+Furthermore it is tested if the header-file includes all necessary includes to compile standalone.
+*/
+
+// API includes
+#include "custom_io/co_sim_EMPIRE_API.h"
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/tests/cpp_tests/test_multiple_includes_api_headers.cpp:22]: With the `static` keyword .. is this still required ?

File: applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+        model_part_io = KM.ModelPartIO(GetFilePath("generic_mdpa_files/Mok_CFD"))
+        model_part_io.ReadModelPart(model_part)
+
+        KM.Logger.GetDefaultOutput().SetSeverity(severity)
+
+        KratosCoSim.EMPIRE_API.EMPIRE_API_sendMesh(model_part)
+
+        # can directly use filecmp because there are no decimal-number issues
+        self.assertTrue(filecmp.cmp(GetFilePath("reference_files/EMPIRE_mesh_For_Sending.vtk_ref"), "EMPIRE_mesh_For_Sending.vtk"))
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py:79]: comparison ... didn't you face any precision issues ?

File: applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+void recvMesh(ModelPart& rModelPart, const std::string rName, const bool UseConditions)
+{
+    KRATOS_ERROR_IF(rModelPart.NumberOfNodes() > 0) << "ModelPart is not empty, it has nodes!" << std::endl;
+    KRATOS_ERROR_IF(rModelPart.NumberOfProperties() > 0) << "ModelPart is not empty, it has properties!" << std::endl;
+    KRATOS_ERROR_IF(rModelPart.IsDistributed()) << "ModelPart cannot be distributed!" << std::endl;
+
+    int numNodes;
+    int numElems;
+    double** nodes = new double*;
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp]: This should be like this 
``` cpp
int** a = new int*[rowCount];
for(int i = 0; i < rowCount; ++i)
    a[i] = new int[colCount];
```

Comment (adityaghantasala) [reply applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp]: https://stackoverflow.com/questions/936687/how-do-i-declare-a-2d-array-in-c-using-new

**Review by adityaghantasala — CHANGES_REQUESTED** [review]
Change the way memory is allocated and deallocated

File: applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py (hunk 7 above)
Comment (philbucher) [reply applications/CoSimulationApplication/tests/test_cosim_EMPIRE_API.py:79]: not in my specific case, but I think in the long run we still have to implement the comparison properly

**Comment by philbucher** [comment]
> Change the way memory is allocated and deallocated

can we please have a look together?
I cannot figure it out...

**Comment by philbucher** [comment]
ok I figured it out, at least I think so
Also it no longer crashes

**Review by adityaghantasala — APPROVED** [review]

**Pull request merged**